Add ZIP 155 (addrv2 message), and update ZIP 252 to reference it (#543)
zcash/zips: zip-0155.rst

::

  ZIP: 155
  Title: addrv2 message
  Owners: Daira Hopwood <daira@electriccoin.co>
  Credits: Wladimir J. van der Laan
           Zancas Wilcox
  Status: Proposed
  Category: Network
  Created: 2021-08-13
  License: BSD-2-Clause
  Discussions-To: <https://github.com/zcash/zips/issues/542>
  Pull-Request: <https://github.com/zcash/zips/pull/543>

Terminology
===========

The key words "MUST", "MUST NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED" and "MAY"
in this document are to be interpreted as described in RFC 2119. [#RFC2119]_

The term "network upgrade" in this document is to be interpreted as described in
ZIP 200. [#zip-0200]_

The terms "Testnet" and "Mainnet" are to be interpreted as described in
section 3.11 of the Zcash Protocol Specification [#protocol-networks]_.

"P2P network" means the Zcash peer-to-peer network.

Abstract
========

This document proposes a new P2P message to gossip longer node addresses over the

> P2P network. This is required to support new-generation Onion addresses, I2P, and

Author: ```suggestion
P2P network. This is required to support new-generation onion addresses, I2P, and
```

potentially other networks that have longer endpoint addresses than fit in the 128
bits of the current ``addr`` message.

Motivation
==========

> Tor v3 hidden services are part of the stable release of Tor since version 0.3.2.9.
> They have various advantages compared to the old hidden services, among which better

Author: ```suggestion
Tor v3 onion services are part of the stable release of Tor since version 0.3.2.9.
They have various advantages compared to the old v2 onion services, among which better
```

encryption and privacy [#Tor-rendezvous-v3]_. These services have 256-bit addresses

> and thus do not fit in the existing ``addr`` message, which encapsulates onion
> addresses in OnionCat IPv6 addresses.

Author: ```suggestion
and thus do not fit in the existing ``addr`` message [#bitcoin-addr]_, which encapsulates
onion addresses in OnionCat IPv6 addresses.
```

Other transport-layer protocols such as I2P have always used longer addresses. This
change would make it possible to gossip such addresses over the P2P network, so that
other peers can connect to them.

Specification
=============

> The ``addrv2`` message is defined as a message where ``pchCommand == "addrv2"``.
> It is  serialized in the standard encoding for P2P messages. Its format is similar
> to the  current ``addr`` message format described in [#Bitcoin-addr-message]_, with
> the difference that the fixed 16-byte IP address is replaced by a network ID and a
> variable-length address, and the services format has been changed to
> [#Bitcoin-CompactSize]_.

Author: ```suggestion
The ``addrv2`` message is defined as a message where the ``command`` field is
(NUL-padded) ``"addrv2"``. It is serialized in the standard encoding for P2P messages.
Its format is similar to the current ``addr`` message format described in
[#Bitcoin-addr-message]_, with the difference that the fixed 16-byte IP address is
replaced by a network ID and a variable-length address, and the services format
has been changed to [#Bitcoin-CompactSize]_.
```

This means that the message contains a serialized vector of the following structure:

  +--------------+-----------------+---------------------+----------------------------------------------------------------+
  | Bytes        | Name            | Data Type           | Description                                                    |
  +--------------+-----------------+---------------------+----------------------------------------------------------------+

>   | 4            | ``time``        | ``uint32``          | Time that this node was last seen as connected to the network. |

Reviewer: `uint32_t`?

  |              |                 |                     | A time in Unix epoch time format.                              |
  +--------------+-----------------+---------------------+----------------------------------------------------------------+
  | ``varies``   | ``services``    | ``CompactSize``     | Service bits. A CompactSize-encoded bit field that is 64 bits  |
  |              |                 |                     | wide.                                                          |
  +--------------+-----------------+---------------------+----------------------------------------------------------------+

>   | 1            | ``networkID``   | ``uint8``           | Network identifier. An 8-bit value that specifies which        |

Reviewer: ![uint8](https://user-images.githubusercontent.com/640955/129425875-efc457da-beaa-41a5-bcf2-f23fd1c4640c.png)

[BIP155](https://github.com/bitcoin/bips/blame/master/bip-0155.mediawiki#L65) specifies `uint8_t` for this field.   I recognize that this is a language specific type. 

I also recognize that `uint8` is not (according to my `clang++ -std=c++17`) a standard type.  

I considered two options:

1.  We could specify that we're implementing against a particular C standard, and use notation derived from that implementation detail.
2.  We could choose type signifiers that are less easily confused with the implementation language syntax.

I am not familiar enough with the ZIP standard to know if it offers guidance.

Author: ``uint8`` is the convention we use generally in Zcash specifications (e.g. [ZIP 225](https://zips.z.cash/zip-0225#transaction-format) or the protocol spec). I'll add it to the Terminology section.

  |              |                 |                     | network is addressed.                                          |
  +--------------+-----------------+---------------------+----------------------------------------------------------------+
  | ``varies``   | ``sizeAddr``    | ``CompactSize``     | The length in bytes of ``addr``.                               |
  +--------------+-----------------+---------------------+----------------------------------------------------------------+
  | ``sizeAddr`` | ``addr``        | ``uint8[sizeAddr]`` | Network address. The interpretation depends on networkID.      |
  +--------------+-----------------+---------------------+----------------------------------------------------------------+
  | 2            | ``port``        | ``uint16``          | Network port. If not relevant for the network this MUST be 0.  |
  +--------------+-----------------+---------------------+----------------------------------------------------------------+

One message can contain up to 1,000 addresses. Clients MUST reject messages with
more addresses.

> Field ``addr`` has a variable length, with a maximum of 512 bytes (4096 bits). Clients

Reviewer: Would it be worthwhile to explicitly rename `addr` to `addrv2`?

Author: That's the name of the message. I think `addr` is fine for the field (as in BIP 155).

> SHOULD reject messages with longer addresses, irrespective of the network ID.

Author: ```suggestion
SHOULD reject messages with a longer ``addr`` field, irrespective of the network ID.
```

The list of reserved network IDs is as follows:

  +------------+-------------+------------------------+-----------------------------------------+
  | Network ID | Enumeration | Address length (bytes) | Description                             |
  +------------+-------------+------------------------+-----------------------------------------+
  | ``0x01``   | ``IPV4``    | 4                      | IPv4 address (globally routed internet) |
  +------------+-------------+------------------------+-----------------------------------------+
  | ``0x02``   | ``IPV6``    | 16                     | IPv6 address (globally routed internet) |
  +------------+-------------+------------------------+-----------------------------------------+

>   | ``0x04``   | ``TORV3``   | 32                     | Tor v3 hidden service address           |

Author: ```suggestion
  | ``0x04``   | ``TORV3``   | 32                     | Tor v3 onion service address            |
```

  +------------+-------------+------------------------+-----------------------------------------+
  | ``0x05``   | ``I2P``     | 32                     | I2P overlay network address             |
  +------------+-------------+------------------------+-----------------------------------------+
  | ``0x06``   | ``CJDNS``   | 16                     | Cjdns overlay network address           |
  +------------+-------------+------------------------+-----------------------------------------+

Network ID ``0x03`` is intentionally not assigned. In BIP 155 [#bip-0155]_ it was

> assigned to Tor v2 hidden service addresses, but those addresses are no longer
> supported by the latest Tor client code, and MUST NOT be used once this ZIP is
> deployed.

Author: ```suggestion
assigned to Tor v2 onion addresses, but those addresses are no longer supported
by the latest Tor client code, and MUST NOT be used once this ZIP is deployed.
```

Clients SHOULD gossip addresses from all known networks even if they are currently
not connected to some of them. That could help multi-homed nodes and make it more

> difficult for an observer to tell which networks a node is connected to.

Reviewer 2: What should clients do about known but un-validated addresses?
(For example, a client that is based on this spec, but doesn't check that tor onion service addresses are valid ed25519 keys.)

The above paragraph suggests that they should gossip them, but the below paragraph says that we don't want to gossip un-validated addresses.

Author: Note to self: check https://github.com/bitcoin/bitcoin/blob/master/src/netaddress.cpp

Reviewer: This comment is suggestive:

https://github.com/bitcoin/bitcoin/blob/92f3a4b4d0531976c9484f5b5075487e0925b619/src/netaddress.cpp#L149

       // TORv2-in-IPv6 (unsupported). Unserialize as !IsValid(), thus ignoring them.
       // Mimic a default-constructed CNetAddr object which is !IsValid() and thus
       // will not be gossiped, but continue reading next addresses from the stream.

I'm examining `IsValid`.

Reviewer: https://github.com/bitcoin/bitcoin/blob/4fc15d15667d9d9c4fb5515ce73c05b4596298ec/src/netaddress.h#L223

Author: I changed this to "Clients SHOULD gossip valid, potentially routable addresses...". It is up to a node implementation which addresses on each network it considers valid and potentially routable.

Clients MUST NOT gossip addresses from unknown networks, because they have no means
to validate those addresses and so can be tricked to gossip invalid addresses.

> Further network ID numbers MUST be reserved in a new ZIP.
> 

Author: ```suggestion
```

Clients MUST reject messages that contain addresses that have a different length
than specified in this table for a specific network ID, as these are meaningless.

Network address encodings
-------------------------

The ``IPV4`` and ``IPV6`` network IDs use addresses encoded in the usual way for
binary IPv4 and IPv6 addresses in network byte order (big endian).

Tor v3 address encoding
'''''''''''''''''''''''

According to the spec [#Tor-rendezvous-v3]_, version 3 ``.onion`` addresses are
encoded as follows::

>   onion_address = base32(PUBKEY | CHECKSUM | VERSION) + ".onion"
>   CHECKSUM = H(".onion checksum" | PUBKEY | VERSION)[:2]

Author: ```suggestion
  onion_address = base32(PUBKEY || CHECKSUM || VERSION) + ".onion"
  CHECKSUM = H(".onion checksum" || PUBKEY || VERSION)[:2]  // first 2 bytes
```

where:

> * ``PUBKEY`` is the 32 bytes Ed25519 master pubkey of the hidden service;

Author: ```suggestion
* ``PUBKEY`` is the 32 bytes Ed25519 master pubkey of the onion service;
```

> * ``VERSION`` is a one byte version field (default value ``0x03``);

Author: ```suggestion
* ``VERSION`` is a one-byte version field (default value ``0x03``);
```

* ``".onion"`` and ``".onion checksum"`` are constant strings;
* ``CHECKSUM`` is truncated to two bytes before inserting it in ``onion_address``;

> * ``H()`` is the SHA3-256 cryptographic hash function.

Author: ```suggestion
* ``H()`` is the SHA3-256 cryptographic hash function. [#NIST-SHA3]_
```

Tor v3 addresses MUST be sent with the ``TORV3`` network ID, with the 32-byte

> ``PUBKEY`` part in the address field. As ``VERSION`` will always be '\x03' in the

Author: ```suggestion
``PUBKEY`` part in the ``addr`` field. As ``VERSION`` will always be '\x03' in the
```

case of v3 addresses, this is enough to reconstruct the onion address.

I2P address encoding
''''''''''''''''''''

Like Tor, I2P naming uses a base32-encoded address format [#I2P-naming]_.

I2P uses 52 characters (256 bits) to represent the full SHA-256 hash, followed by

> ``.b32.i2p``.

Reviewer: Should we mention padding bits?  52 base32 characters is 52*5 = 260 bits, so there's 4 bits of extra information that aren't part of the SHA-256 hash's representation.

Author: Padding is part of the base32 spec. I think this is clear enough.

Author: ```suggestion
``.b32.i2p``. The base32 encoding does not include ``"="`` padding characters.
```

I2P addresses MUST be sent with the ``I2P`` network ID, with the decoded SHA-256 hash
as address field.

Cjdns address encoding
''''''''''''''''''''''

Cjdns addresses are simply IPv6 addresses in the ``fc00::/8`` range

> [#Cjdns-whitepaper]_. They MUST be sent with the ``CJDNS`` network ID.

Author: ```suggestion
[#Cjdns-whitepaper]_. They MUST be sent with the ``CJDNS`` network ID.
They are encoded in network byte order (big endian) as for the ``IPV6`` network ID.
```

Deployment
----------

Support for this specification is signalled by peer protocol version 170015 (on
both Testnet and Mainnet). Note that this is the same peer protocol version that
signals support for NU5 on Mainnet [#zip-0252]_.

Nodes that have not negotiated peer protocol version 170015 or higher on a given
connection, MUST NOT send ``addrv2`` messages on that connection.

A node that has negotiated peer protocol version 170015 or higher on a given
connection, MAY still send ``addr`` messages on the connection, and MUST handle
received ``addr`` messages as it would have done prior to this ZIP.

Rationale
=========

This ZIP is closely based on BIP 155 [#bip-0155]_, with the following changes:

* Deployment: support for the ``addrv2`` message is signalled by advertising a

>   peer-to-peer protocol version of 170015 or higher, not by sending a ``sendaddrv2``

Author: ```suggestion
  peer protocol version of 170015 or higher, not by sending a ``sendaddrv2``
```

  message. This is motivated by a desire to avoid an exponential explosion in the
  space of possible feature configurations in a given peer-to-peer connection. In
  Zcash, unlike Bitcoin, the space of such configurations is effectively constant no
  matter how many peer-to-peer protocol changes are made, because nodes that do not
  support a given peer protocol version will drop off the network over time if they do
  not support the latest Network Upgrade. The feature configuration for a given
  connection is also established at version negotiation, and cannot change after that
  point without reconnecting. Other peer-to-peer protocol changes ported from the
  Bitcoin peer-to-peer protocol, for example the ``MSG_WTX`` inv type defined in
  ZIP 239 [#zip-0239]_, have taken the same approach to signalling.

> * No Network ID for Tor v2 hidden service addresses: the Tor network is expected to
>   have removed support for these addresses in the timeframe for deployment of this
>   ZIP.

Author: ```suggestion
* No Network ID for Tor v2 onion addresses: the Tor network is expected to have
  removed support for these addresses in the timeframe for deployment of this ZIP.
```

* Clients MUST, rather than SHOULD, reject ``addrv2`` messages with more than 1,000
  addresses. Making this a consistent requirement promotes interoperability.

* Clients MUST NOT, rather than SHOULD NOT, gossip addresses from unknown networks.

* Clients MUST, rather than SHOULD, reject messages that contain addresses that have

>   a different length than specified for a specific network ID.

Author: ```suggestion
  a different length than specified for a specific network ID, or a length greater
  than the 512-byte maximum for an unknown ID.
```

Acknowledgements for BIP 155:

* Jonas Schnelli: change ``services`` field to ``CompactSize``, to make the message
  more compact in the likely case instead of always using 8 bytes.
* Gregory Maxwell: various suggestions regarding extensibility.

References
==========

.. [#RFC2119] `RFC 2119: Key words for use in RFCs to Indicate Requirement Levels <https://www.rfc-editor.org/rfc/rfc2119.html>`_
.. [#protocol-networks] `Zcash Protocol Specification, Version 2020.2.14 [NU5 proposal]. Section 3.12 Mainnet and Testnet <protocol/protocol.pdf#networks>`_
.. [#bip-0155] `BIP 155: addrv2 message <https://github.com/bitcoin/bips/blob/master/bip-0155.mediawiki>`_
.. [#zip-0200] `ZIP 200: Network Upgrade Mechanism <zip-0200.rst>`_
.. [#zip-0239] `ZIP 239: Relay of Version 5 Transactions <zip-0239.rst>`_
.. [#zip-0252] `ZIP 252: Deployment of the NU5 Network Upgrade <zip-0252.rst>`_
.. [#Bitcoin-addr-message] `Protocol documentation: addr. Bitcoin Wiki <https://en.bitcoin.it/wiki/Protocol_documentation#addr>`_
.. [#Bitcoin-CompactSize] `Variable length integer. Bitcoin Wiki <https://en.bitcoin.it/wiki/Protocol_documentation#Variable_length_integer>`_

> .. [#Tor-rendezvous-v3] `Tor Rendezvous Specification - Version 3 <https://gitweb.torproject.org/torspec.git/tree/rend-spec-v3.txt>`_

Author: ```suggestion
.. [#Tor-rendezvous-v3] `Tor Rendezvous Specification - Version 3 <https://gitweb.torproject.org/torspec.git/tree/rend-spec-v3.txt>`_
.. [#NIST-SHA3] `NIST FIPS 202 - SHA-3 Standard: Permutation-Based Hash and Extendable-Output Functions <https://csrc.nist.gov/publications/detail/fips/202/final>`_
```
